File: rotatelist.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode rotateRight(ListNode head, int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int length = 0;
        ListNode p = head;
        ListNode tail = head;
        while(p!=null){
            length++;
            tail = p;
            p = p.next;
        }
        if(length==0){
            return head;
        }
        n = n%length;
        if(n==0) return head;
        n = length-n;
        n = n-1;
        p = head;
        while(n>0){
            p = p.next;
            n--;
        }
        ListNode newhead = p.next;
        tail.next = head;
        p.next = null;
        return newhead;
    }
}

File: cpp/RotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        if(head==NULL) {
            return NULL;
        }
        int n = 0;
        
        ListNode* tail = head;
        while(tail->next!=NULL) {
            n++;
            tail = tail->next;
        }
        n++;
        k = k%n;
        if(k==0) {
            return head;
        }
        int t = n-k-1;
        ListNode* p = head;
        while(t--) {
            p = p->next;
        }
        ListNode* headNew = p->next;
        tail->next = head;
        p->next = NULL;
        return headNew;
    }
};
